### ISSUE TITLE
Associate Dialog Header with Dialog Title via aria-labelledby

### ISSUE THREAD

**Issue by scurker** [dequelabs/cauldron#1259]
[Current aria practices suggest](https://www.w3.org/WAI/ARIA/apg/patterns/dialog-modal/examples/dialog/) that we should be using `aria-labelledby` to associate the accessible name of the dialog with the title of the dialog. This is currently raised in `axe-core` as a serious best practice issue via [`aria-dialog-name`](https://dequeuniversity.com/rules/axe/4.7/aria-dialog-name).

All of our dialogs currently have a heading, so the needed changes to satisfy this are:

- Auto-generate an id with `react-id-generator` and set the id on the heading element
- Associate the `role="dialog"` container with an `aria-label` with the id reference of the heading
